File: sources/java/10_pi/Main.java
// task 10 pi — expected output: 44889
// build: javac Main.java    run: java Main
// build (graalvm native-image): native-image -O2 Main    run: ./Main
// Gibbons' unbounded spigot on java.math.BigInteger; the leading 3 counts as one of the
// 10000 digits, and only the digit sum is printed.

import java.math.BigInteger;

public class Main {
    private static final BigInteger TWO = BigInteger.valueOf(2);
    private static final BigInteger THREE = BigInteger.valueOf(3);
    private static final BigInteger FOUR = BigInteger.valueOf(4);
    private static final BigInteger SEVEN = BigInteger.valueOf(7);
    private static final BigInteger TEN = BigInteger.TEN;

    public static void main(String[] args) {
        BigInteger q = BigInteger.ONE;
        BigInteger r = BigInteger.ZERO;
        BigInteger t = BigInteger.ONE;
        BigInteger k = BigInteger.ONE;
        BigInteger n = THREE;
        BigInteger l = THREE;

        long digitSum = 0;
        int digits = 0;
        while (digits < 10000) {
            if (q.multiply(FOUR).add(r).subtract(t).compareTo(n.multiply(t)) < 0) {
                digitSum += n.intValue();
                digits += 1;
                BigInteger qNext = q.multiply(TEN);
                BigInteger rNext = r.subtract(n.multiply(t)).multiply(TEN);
                n = q.multiply(THREE).add(r).multiply(TEN).divide(t).subtract(n.multiply(TEN));
                q = qNext;
                r = rNext;
                // t, k and l are unchanged on this branch
            } else {
                BigInteger qNext = q.multiply(k);
                BigInteger rNext = q.multiply(TWO).add(r).multiply(l);
                BigInteger tNext = t.multiply(l);
                n = q.multiply(k.multiply(SEVEN).add(TWO)).add(r.multiply(l)).divide(tNext);
                q = qNext;
                r = rNext;
                t = tNext;
                k = k.add(BigInteger.ONE);
                l = l.add(TWO);
            }
        }
        System.out.println(digitSum);
    }
}

File: sources/fsharp/10_pi/main.fs
// task 10 pi — expected output: 44889
// build: dotnet build -c Release    run: dotnet run

open System.Numerics

// Gibbons' unbounded spigot, on BigInteger. Emits the leading 3 first,
// then one decimal digit of pi at a time. Only the digit sum is printed.
let zero = BigInteger.Zero
let one = BigInteger.One
let two = BigInteger(2)
let three = BigInteger(3)
let four = BigInteger(4)
let seven = BigInteger(7)
let ten = BigInteger(10)

let mutable q = one
let mutable r = zero
let mutable t = one
let mutable k = one
let mutable n = three
let mutable l = three

let mutable emitted = 0
let mutable sum = 0

while emitted < 10000 do
    if four * q + r - t < n * t then
        sum <- sum + int n
        emitted <- emitted + 1
        if emitted < 10000 then
            let nNext = (ten * (three * q + r)) / t - ten * n
            q <- ten * q
            r <- ten * (r - n * t)
            n <- nNext
    else
        let qNext = q * k
        let rNext = (two * q + r) * l
        let tNext = t * l
        let kNext = k + one
        let nNext = (q * (seven * k + two) + r * l) / (t * l)
        q <- qNext
        r <- rNext
        t <- tNext
        k <- kNext
        n <- nNext
        l <- l + two

printfn "%d" sum
